src/nvmeof/NVMeofGwMonitorClient.cc

```diff
@@ -0,0 +1,394 @@
+  client_messenger->shutdown();
+}
+
+void NVMeofGwMonitorClient::handle_nvmeof_gw_map(ceph::ref_t<MNVMeofGwMap> nmap)
+{
+  auto &new_map = nmap->get_map();
+  auto group_key = std::make_pair(pool, group);
+  dout(0) << "handle nvmeof gw map: " << new_map << dendl;
+
+  NvmeGwState old_gw_state;
+  auto got_old_gw_state = get_gw_state("old map", map, group_key, name, old_gw_state); 
+  NvmeGwState new_gw_state;
+  auto got_new_gw_state = get_gw_state("new map", new_map, group_key, name, new_gw_state); 
+
+  if (!got_old_gw_state) {
+    if (!got_new_gw_state) {
+      dout(0) << "Can not find new gw state" << dendl;
+      return;
+    }
+    bool set_group_id = false;
+    while (!set_group_id) {
+      NVMeofGwMonitorGroupClient monitor_group_client(
+          grpc::CreateChannel(monitor_address, grpc::InsecureChannelCredentials()));
+      dout(0) << "GRPC set_group_id: " <<  new_gw_state.group_id << dendl;
+      set_group_id = monitor_group_client.set_group_id( new_gw_state.group_id);
+      if (!set_group_id) {
+	      dout(0) << "GRPC set_group_id failed" << dendl;
+	      usleep(1000); // TODO: conf options
+      }
+    }
+  }
+
+  // Make sure we do not get out of order state changes from the monitor
+  if (got_old_gw_state && got_new_gw_state) {
+    dout(0) << "got_old_gw_state: " << old_gw_state << "got_new_gw_state: " << new_gw_state << dendl;
+    ceph_assert(new_gw_state.gw_map_epoch >= old_gw_state.gw_map_epoch);
+  }
+
+  // Gather all state changes
+  ana_info ai;
+  for (const auto& nqn_state_pair: new_gw_state.subsystems) {
+    auto& sub = nqn_state_pair.second;
+    const auto& nqn = nqn_state_pair.first;
+    nqn_ana_states nas;
+    nas.set_nqn(nqn);
+    for (NvmeAnaGrpId  ana_grp_index = 0; ana_grp_index < sub.ana_state.size(); ana_grp_index++) {
+      const auto& old_nqn_state_pair = old_gw_state.subsystems.find(nqn);
+      auto found_old_nqn_state = (old_nqn_state_pair != old_gw_state.subsystems.end());
+      auto new_group_state = sub.ana_state[ana_grp_index];
+
+      // if no state change detected for this nqn, group id
+      if (got_old_gw_state && found_old_nqn_state &&
+           new_group_state == old_nqn_state_pair->second.ana_state[ana_grp_index]) {
+        continue;
+      }
+      ana_group_state gs;
+      gs.set_grp_id(ana_grp_index + 1); // offset by 1, index 0 is ANAGRP1
+      const auto& new_agroup_state = new_group_state.first;
+      const epoch_t& blocklist_epoch = new_group_state.second;
+
+      if (new_agroup_state == GW_EXPORTED_STATES_PER_AGROUP_E::GW_EXPORTED_OPTIMIZED_STATE &&
+          blocklist_epoch != 0) {
+        // Check if we need to wait for a newer OSD map before starting
+        dout(0) << "Check if ready for blocklist osd map epoch: " << blocklist_epoch << dendl;
+        std::set<entity_addr_t> newly_blocklisted;
+        objecter.consume_blocklist_events(&newly_blocklisted);
+        dout(0) << "Consumed new blocklists: " << newly_blocklisted << dendl;
```

Comment:
`consume_blocklist_events()` is called to make sure the objecter consumes new blocklist events, the value returned is currently for logging only, since the NVMeofGwMonitorClient's last_osd_epoch traces traces only advances of the blocklisted epochs received from the monitor, see line 344-347.

Reply:
consume_blocklist_events is a service the Objecter provides out to other users — in the existing code base, it is used only by the MDS, to make sure the MDS sees every blocklisted client IP (to do its own blocklist processing on top). It's not something you need to invoke unless you're using it, and in fact unless you set Objecter::blocklist_events_enabled=true, this set will always be empty.

Reply:
@gregsfortytwo consume_blocklist_events() code really helped out when we were figuring out the blocklist logic, but it might not be needed at the moment, as long as the `objecter.with_osdmap()` line 335 keeps doing its thing, we're golden. created [wip](https://github.com/ceph/ceph-ci/tree/wip-baum-20240313-00) to test to the [diff](https://github.com/ceph/ceph-ci/commit/7dda1050a2065217d18902672915ac06dd460a80). Once we a shaman build is available I could verify the change in ceph-nvmeof HA CI, then I'll integrate this change into this PR. Meanwhile I am still waiting for the current PR version [build](https://shaman.ceph.com/builds/ceph/ceph-nvmeof-mon/d4ecff0fa053358bcc58ea38e914fe80293f6686/) : Been waiting for "3 hours, 59 minutes, 9 seconds" as per Shaman, and still no sign of any builds kicking off. crossing my fingers for some feedback by tomorrow 😊